### PR TITLE
feat: add cascade toggle logic to enterprise sub inclusion boolean

### PR DESCRIPTION
**Pull request by alex-sheehan-edx**
https://2u-internal.atlassian.net/jira/software/c/projects/ENT/boards/630?modal=detail&selectedIssue=ENT-5840

When an org subscription inclusion is toggled, all courses, course runs and programs included in the org should check if they should be toggled.

accounts for course/program types as well.

### CODE REVIEW

File: course_discovery/apps/course_metadata/models.py [hunk 1]
```diff
@@ -2724,8 +2743,6 @@ def is_active(self):
         return self.status == ProgramStatus.Active
 
     def _check_enterprise_subscription_inclusion(self):
-        if self.type == 'micromasters':
-            return False
```
Comment (kiram15) [inline course_discovery/apps/course_metadata/models.py:2769]: TY!

Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/models.py:2769]: of course!

File: course_discovery/apps/course_metadata/models.py [hunk 2]
```diff
@@ -992,6 +1008,9 @@ def save(self, *args, **kwargs):
         kwargs['force_insert'] = False
         kwargs['force_update'] = True
         super().save(*args, **kwargs)
```
Comment (kiram15) [inline course_discovery/apps/course_metadata/models.py:1038]: In the above method of `_check_enterprise_subscription_inclusion`, we now need to check if it is of the course types we allow. Instead of excluding, we should do an include so that if new weird ones are added, we don't have to change the code. Using the CourseTypes detailed in this file, this would probably look something like this.

```
enterprise_catalog_course_types = [CourseType.AUDIT, CourseType.VERIFIED_AUDIT, CourseType.PROFESSIONAL, 
CourseType.CREDIT_VERIFIED_AUDIT, CourseType.EMPTY] (probably in a constants file)

if self.type in enterprise_catalog_course_types or self.enterprise_subscription_inclusion is False:
...
```




.

File: course_discovery/apps/course_metadata/models.py (hunk 1 above)
Comment (kiram15) [reply course_discovery/apps/course_metadata/models.py:2769]: Actually, replace this with

```
enterprise_catalog_program_types = [ProgramType.XSERIES, ProgramType.MICROMASTERS, ProgramType.PROFESSIONAL_CERTIFICATE, ProgramType.PROFESSIONAL_PROGRAM_WL, ProgramType.LICENSE, ProgramType.MICROBACHELORS]

#We exclude Bachelors, Masters, and Doctorate programs as the cost per user would be too high
if self.type not in enterprise_catalog_program_types:
  return false;
```

**Review by kiram15 — COMMENTED** [review]
So we're hacking out some changed product requirements but I tried to type as much info as possible to give you context!

File: course_discovery/apps/course_metadata/models.py [hunk 3]
```diff
@@ -267,8 +267,24 @@ def save(self, *args, **kwargs):
         We cache the key here before saving the record so that we can hit the correct
         endpoint in lms.
         """
+        old_copy = Organization.objects.filter(pk=self.pk).first()
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: pk is unique, hence you can use `Organization.objects.get(pk=self.pk)`

File: course_discovery/apps/course_metadata/models.py [hunk 4]
```diff
@@ -267,8 +267,24 @@ def save(self, *args, **kwargs):
...
+        if toggle_course_subscription_inclusion:
+            courses = Course.objects.filter(authoring_organizations__pk=self.pk)
+            for course in courses:
+                course.enterprise_subscription_inclusion = self.enterprise_subscription_inclusion
+                course.save()
+
+            programs = Program.objects.filter(courses__in=courses)
+            for program in programs:
+                program.save()
+
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py]: This can be an expensive operation. There are post_save signal handles for Course model that then call other services for updating information. I would advocate making this part an async task. Also, keep draft vs non-draft distinction and impact in mind when doing this operation.

Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/models.py]: thanks for the suggestion!

Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/models.py]: @DawoudSheraz could you talk a little more about draft vs non-draft implications? Theoretically we'd want all courses associated with the org to be updated when the org is updated, no? I could potentially only save if there's a change in the `enterprise_subscription_inclusion` value but I'm not sure that's what you meant.

File: course_discovery/apps/course_metadata/models.py [hunk 5]
```diff
@@ -2731,8 +2748,10 @@ def is_active(self):
         return self.status == ProgramStatus.Active
 
     def _check_enterprise_subscription_inclusion(self):
-        if self.type == 'micromasters':
+        # We exclude Bachelors, Masters, and Doctorate programs as the cost per user would be too high
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/models.py:2787]: Shouldn't we exclude `LICENSE` as well?

File: course_discovery/apps/course_metadata/tasks.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    """
+    courses = Course.objects.filter(
+        authoring_organizations__pk=org_pk, draft=False, enterprise_subscription_inclusion__in=[None, True],
+    )
+    course_ids = []
+    for course in courses:
+        course.enterprise_subscription_inclusion = org_sub_inclusion
+        course.save()
+        course_ids.append(course.id)
+    programs = Program.objects.filter(courses__in=course_ids)
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/tasks.py]: are we interested in 2U programs which does not have associated courses?

File: course_discovery/apps/course_metadata/tasks.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+@shared_task()
+def update_org_program_and_courses(org_pk, org_sub_inclusion):
+    """
+    Task to update course skills.
+    Arguments:
+        org_pk (int): primary key of the organization
+        org_sub_inclusion (bool): whether or not the org is included in enterprise subscriptions
+    """
+    courses = Course.objects.filter(
+        authoring_organizations__pk=org_pk, draft=False, enterprise_subscription_inclusion__in=[None, True],
```
Comment (azanbinzahid) [inline course_discovery/apps/course_metadata/tasks.py]: Can it have a `False` value? What if the value is prev set to `False` and we need to set it `True`?

File: course_discovery/apps/course_metadata/models.py (hunk 5 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/models.py:2787]: will need @kiram15's feedback on this.

File: course_discovery/apps/course_metadata/tasks.py (hunk 6 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/tasks.py]: I don't believe so because those associated programs will default to false and won't need updating in this post org save logic. Basically we only want to toggle the subscription inclusion of programs that have OCM courses that belong to the org.

File: course_discovery/apps/course_metadata/tasks.py (hunk 7 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/tasks.py]: No- any course that is set to `False` we want to stay false. We only want to toggle courses that were previously `True` or were never set in the first place (all new courses will default have a value so these `None` cases only apply to courses generated before the sub inclusion logic was implemented.

File: course_discovery/apps/course_metadata/models.py (hunk 5 above)
Comment (kiram15) [reply course_discovery/apps/course_metadata/models.py:2787]: Do we have any documentation on why License should be excluded? In talking with our PM to go over requirements that wasn't one that was mentioned.

File: course_discovery/apps/course_metadata/tests/test_models.py [hunk 8]
```diff
@@ -2033,18 +2126,17 @@ def test_staff_no_advertised_course_run(self):
...
+        program1 = factories.ProgramFactory(type=program_type, courses=course_list)
         assert program1.enterprise_subscription_inclusion is False
-
-        course3 = factories.CourseFactory(enterprise_subscription_inclusion=True)
-        course4 = factories.CourseFactory(enterprise_subscription_inclusion=True)
+        course3 = factories.CourseFactory(enterprise_subscription_inclusion=True, type=course_type)
+        course4 = factories.CourseFactory(enterprise_subscription_inclusion=True, type=course_type)
         course_list2 = [course3, course4]
-        program2 = factories.ProgramFactory(type=type1, courses=course_list2)
+        program2 = factories.ProgramFactory(courses=course_list2, type=program_type)
```
Comment (kiram15) [inline course_discovery/apps/course_metadata/tests/test_models.py:2140]: Can we add in a test that checks against a program that is not included in the `is_enterprise_catalog_program_type` list?

Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/tests/test_models.py:2140]: yeah definitely

File: course_discovery/settings/test.py [hunk 9]
```diff
@@ -72,7 +72,8 @@
 
 CELERY_TASK_IGNORE_RESULT = True
 
-CELERY_RESULT_BACKEND = f'file://{tempfile.TemporaryDirectory()}'
+results_dir = tempfile.TemporaryDirectory()
```
Comment (alex-sheehan-edx) [inline course_discovery/settings/test.py:75]: seems like testing celery needed a small tweak

File: course_discovery/apps/course_metadata/tasks.py [hunk 10]
```diff
@@ -0,0 +1,31 @@
+"""
+Celery tasks for course metadata.
+"""
+import logging
+
+from celery import shared_task
+
+from .models import Course, Program
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tasks.py]: nit: please use absolute import path.

File: course_discovery/apps/course_metadata/tasks.py [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+
+LOGGER = logging.getLogger(__name__)
+
+
+@shared_task()
+def update_org_program_and_courses(org_pk, org_sub_inclusion):
+    """
+    Task to update course skills.
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tasks.py]: Please rename the task to showcase what is actually being done. Same for the docstrings.

File: course_discovery/apps/course_metadata/tasks.py (hunk 7 above)
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tasks.py]: - Add an additional filter on course type, to filter only those courses whose types can be included in the enterprise catalog
- I suspect draft=False would not work here. What we should do is to filter the draft versions and if a draft version has an official version (non-draft), copy the changes from the draft to the non-draft version. This will make sure to keep draft and non-draft versions in sync. Otherwise, whenever any of the courses are edited by this task, their next edit run via the publisher, which performs edits on draft first will override these changes in DB when copying values from Draft -> Non-draft. Let me know if you need further clarification on it.

File: course_discovery/apps/course_metadata/tasks.py (hunk 6 above)
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tasks.py]: Add filtering against allowed program types here as well.

File: course_discovery/apps/course_metadata/signals.py [hunk 12]
```diff
@@ -180,6 +182,14 @@ def update_or_create_salesforce_organization(instance, created, **kwargs):
             util.update_publisher_organization(instance)
 
 
+@shared_task
+@receiver(post_save, sender=Organization)
+def hydrate_org_programs_and_courses(instance, created, **kwargs):  # pylint: disable=unused-argument
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/signals.py]: nit: this can be renamed to a meaningful name.

File: course_discovery/apps/course_metadata/models.py [hunk 13]
```diff
@@ -992,6 +1005,9 @@ def save(self, *args, **kwargs):
         kwargs['force_insert'] = False
         kwargs['force_update'] = True
         super().save(*args, **kwargs)
+        course_runs = CourseRun.objects.filter(course=self)
+        for course_run in course_runs:
+            course_run.save()
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py:1044]: Why is this needed?

Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/models.py:1044]: The course run subscription tagging status is calculated and updated when the course run is saved. This is determined based on the ent sub tagging status of it's parent course. When a course is saved, we want to update the tagging status of each of its course runs.

File: course_discovery/apps/course_metadata/signals.py (hunk 12 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/signals.py]: What were you thinking? To me this explains exactly what's going on. We have a parent org that has been updated and we need to ensure the children programs and courses of the are up to date with subscription tagging.

File: course_discovery/apps/course_metadata/tasks.py (hunk 11 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/tasks.py]: oops! my mistake for missing the doc string.

File: course_discovery/apps/course_metadata/tasks.py (hunk 6 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/tasks.py]: I'm not sure that's necessary. We already have allowed program type filter on the program save. Subscription inclusion for programs is calculated on save same as course runs and does the filtering for us for allowed types.

File: course_discovery/apps/course_metadata/tasks.py (hunk 7 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/tasks.py]: > Add an additional filter on course type, to filter only those courses whose types can be included in the enterprise catalog

this should be taken care of by the course's save `_check_enterprise_subscription_inclusion` logic

Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/tasks.py]: > I suspect draft=False would not work here. What we should do is to filter the draft versions and if a draft version has an official version (non-draft), copy the changes from the draft to the non-draft version. This will make sure to keep draft and non-draft versions in sync. Otherwise, whenever any of the courses are edited by this task, their next edit run via the publisher, which performs edits on draft first will override these changes in DB when copying values from Draft -> Non-draft. Let me know if you need further clarification on it.

I think I need more clarifications on this. Would you be available to huddle and talk at some point?

File: course_discovery/apps/course_metadata/signals.py (hunk 12 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/signals.py]: Something along the lines of `update_enterprise_inclusion_for_courses_and_programs`.

File: course_discovery/apps/course_metadata/tasks.py (hunk 7 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/tasks.py]: Adding the filter on course type will only bring those results whose enterprise inclusion value needs to be updated.

File: course_discovery/apps/course_metadata/models.py [hunk 14]
```diff
@@ -992,6 +1005,9 @@ def save(self, *args, **kwargs):
         kwargs['force_insert'] = False
         kwargs['force_update'] = True
         super().save(*args, **kwargs)
+        course_runs = CourseRun.objects.filter(course=self)
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/models.py:1042]: nit: a small comment on why this is done will be helpful for the future devs.

File: course_discovery/apps/course_metadata/models.py (hunk 5 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/models.py:2787]: License and Bachelors are new types added for new degrees import. Maybe they can be added here in a followup PR once confirmed internally.

File: course_discovery/apps/course_metadata/tests/test_models.py [hunk 15]
```diff
@@ -1238,6 +1245,91 @@ def test_user_organizations(self):
 
         assert len(Organization.user_organizations(user)) == 1
 
+    def test_org_enterprise_subscription_inclusion_toggle_course(self):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tests/test_models.py:1249]: 👍🏽 Nicely done.

**Review by DawoudSheraz — APPROVED** [review]

File: course_discovery/apps/course_metadata/models.py (hunk 5 above)
Comment (azanbinzahid) [reply course_discovery/apps/course_metadata/models.py:2787]: In essence, these were added together: https://github.com/openedx/course-discovery/pull/3450

File: course_discovery/apps/course_metadata/models.py (hunk 14 above)
Comment (alex-sheehan-edx) [reply course_discovery/apps/course_metadata/models.py:1042]: happy to add a comment!
